### ISSUE TITLE
Image perspective transformation

### ISSUE THREAD

**Issue by guillevc** [lovell/sharp#2095]
Is there any way to process the image transforming its perspective using this library?

This is what I mean:
![example](https://i.stack.imgur.com/ARqWb.png)

If not, do you know any approach to make this?
Is this a potential feature for Sharp?

Cheers!

**Comment by lovell** [comment]
Hi, this would require exposing the [vips_affine](https://libvips.github.io/libvips/API/current/libvips-resample.html#vips-affine) operation in sharp. Doing so was previously discussed at #998 but that issue resulted in the exposure of only the `vips_rotate` wrapper (which is a "helper" around `vips_affine`).

Happy to accept a PR for this if you're able.

**Comment by guillevc** [comment]
Does libvips allow us to make a projective transformation though?
![Projective transformation](https://www.graphicsmill.com/docs/gm/TransformationsDifference.png)

I'm guessing there is no way to do this with `vips_affine`.

Still, I'll give it a go and implement a bridge to `vips_affine`.

**Comment by Xaekai** [comment]
I believe vips_affine should be exposed.

My use case is simple: I've gotten used to using nohalo/lohalo after it became available in GIMP, and like the results it produces. I want to create a simple emote image server to my esoteric and very whimsical specifications. All upscales will be produced by the waifu2x service that runs on the same host. And all downscaling to requested size will be done with nohalo.

If that means writing my own calls to libvips, that's fine. But I'd like to not have to create my own bindings when Sharp already exists.

**Comment by guillevc** [comment]
I'll make a PR soon exposing `vips_affine`

**Comment by metaskills** [comment]
Please and post the link here so I can follow along!

**Comment by guillevc** [comment]
@metaskills here's the PR #2320

**Comment by banli17** [comment]
affine does not seem to be able to do perspective deformation, such as a square into a trapezoid, imagemagick can do it through commands，is it？

```
magick convert -virtual-pixel transparent -distort perspective "0,0,0,8 0,390,132,323 558,390,558,265 558,0,400,0" 1.png 2.png
```

**Comment by GavinJoyce** [comment]
@guillevc did you ever manage to achieve projective transformation?

I've been searching for ways of achieving it with libvips, this seems to be an approach using `mapim`: https://stackoverflow.com/a/50219715

**Comment by guillevc** [comment]
@GavinJoyce Nope, I ended up using OpenCV with Python for projective transformation

**Comment by mariusa** [comment]
What would be the next steps to achieve projective transformation? `vips_affine` is exposed now, but doesn't seem to help.

**Comment by lovell** [comment]
As I said in #3071, it might be possible to achieve this with some of the new `transform` features of SVG, but you'll have to wait for #2884

**Comment by zatakiya** [comment]
[This](https://github.com/alxcube/lens) library supports perspective transform with pure javascript. Might be a good place to take reference to add perspective transformation feature in sharp.

**Comment by paintoshi** [comment]
> [This](https://github.com/alxcube/lens) library supports perspective transform with pure javascript. Might be a good place to take reference to add perspective transformation feature in sharp.

Thanks for this! Worked as a charm as I combined with sharp to auto-generate distorted images from input+mask in a node application.

Mask:
![marketing_2](https://github.com/user-attachments/assets/8051f2ef-0aa0-48a7-b181-10437f19ccee)

Result:
![image](https://github.com/user-attachments/assets/918416d9-8aae-4d6e-bdbf-296bf0aa3af9)

![image](https://github.com/user-attachments/assets/7606f0db-a5d5-48b6-ae5b-d5a3d0aaa023)

**Comment by SanderVanLeeuwen** [comment]
The library mentioned above is a good (temporary) solution for 4-point transforms. For example, it helps us transform images in such a way we can create a composite out of a base template (room) and an artwork. With some fiddling to calculate the proper perspective transform, the results are really nice:

![Image](https://github.com/user-attachments/assets/ef44a2a7-abff-4794-b533-d1ae7eb2af2e)

composite result:

![Image](https://github.com/user-attachments/assets/2160b930-2581-432e-a53b-f8b6d1c381f6)

There's one downside; it's slow. Not to badmouth the lens library, on the contrary, but the Sharp library is so fast in what it does that nothing really can compare. E.g., we're talking about 2 or 3 seconds here compared to all the combined Sharp actions which only take 200 to 300 ms.

From what I understand, a native implementation relies on https://github.com/lovell/sharp/issues/2884 which in it's turn waits for https://gitlab.gnome.org/GNOME/librsvg/-/issues/754. Both are 3 to 4 year old issues / feature requests. Is there a realistic time path where 4-point transforms can be achieved through this library either by using SVG's or natively, or is that totally dependent on who's willing to implement it?

I would love to contribute to this, but unfortunately I'm really lacking the knowledge here to be of help from a coding point of view.

**Comment by jcupitt** [comment]
Hello, there's an example using libvips for perspective transforms in this old stackoverflow post:

https://stackoverflow.com/a/50219715/894763

You make an index image with `xyz`, then transform that with the equation you need to make an index image, then use `mapim` to remap the source.

https://www.libvips.org/API/current/method.Image.mapim.html

Copy-pasting the code for convenience:

```ruby
require 'vips'

image = Vips::Image.new_from_file ARGV[0]

# perspective distortion: each pixel (x', y') in the output image is
# interpolated from pixel (x, y) in the input using:
#
#   x' = (A x + B y + C) / (G x + H y + 1)
#   y' = (D x + E y + F) / (G x + H y + 1)
#
# where the constants A .. H are from the transform matrix T
#
# T = [A, B, .. H]

T = [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0003, 0.0001]

# make an index image where pixels have the value of their (x, y) coordinates
i = Vips::Image.xyz image.width, image.height

x = (i[0] * T[0] + i[1] * T[1] + T[2]) / (i[0] * T[6] + i[1] * T[7] + 1)
y = (i[0] * T[3] + i[1] * T[4] + T[5]) / (i[0] * T[6] + i[1] * T[7] + 1)

# join up x and y as a map image
m = x.bandjoin y

# and use it to transform our original image
image = image.mapim m 

image.write_to_file ARGV[1]
```

It looks awkward, but it's actually pretty quick. On this PC and with a 6k x 4k source image I see:

```
$ time ./perspective.rb ~/pics/theo.jpg x.jpg

real	0m0.371s
user	0m1.577s
sys	0m0.258s
```

So under 400ms for the whole thing, including all image IO and ruby-vips startup and shutdown.

I've no idea how this could be exposed in sharp though! Perhaps a dedicated `perspective` transform?
